File: src/main/java/com/qtech/utilities/datetime/Date.java
package com.qtech.utilities.datetime;

import java.io.Serializable;
import java.time.DayOfWeek;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.chrono.IsoChronology;
import java.time.chrono.IsoEra;
import java.util.Objects;

@SuppressWarnings("unused")
public class Date implements Serializable {
    private int day;
    private int month;
    private int year;

    public static Date current() {
        LocalDateTime dateTime = LocalDateTime.now();
        int day = dateTime.getDayOfMonth();
        int month = dateTime.getMonthValue();
        int year = dateTime.getYear();

        return new Date(day, month, year);
    }

    public Date(int day, int month, int year) {
        this.day = day;
        this.month = month;
        this.year = year;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Date date = (Date) o;
        return day == date.day &&
                month == date.month &&
                year == date.year;
    }

    public boolean equalsIgnoreYear(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Date date = (Date) o;
        return day == date.day &&
                month == date.month;
    }

    @Override
    public int hashCode() {
        return Objects.hash(day, month, year);
    }

    public int getDay() {
        return day;
    }

    public void setDay(int day) {
        this.day = day;
    }

    public int getMonth() {
        return month;
    }

    public void setMonth(int month) {
        this.month = month;
    }

    public int getYear() {
        return year;
    }

    public void setYear(int year) {
        this.year = year;
    }

    public DayOfWeek getDayOfWeek() {
        LocalDate localDate = getLocalDate();
        return localDate.getDayOfWeek();
    }

    public int getDayOfYear() {
        LocalDate localDate = getLocalDate();
        return localDate.getDayOfYear();
    }

    public IsoEra getEra() {
        LocalDate localDate = getLocalDate();
        return localDate.getEra();
    }

    public IsoChronology getChronology() {
        LocalDate localDate = getLocalDate();
        return localDate.getChronology();
    }

    @SuppressWarnings("ResultOfMethodCallIgnored")
    public boolean isLeapYear() {
        LocalDate localDate = getLocalDate();
        localDate.toString();
        return localDate.getChronology().isLeapYear(getYear());
    }

    private LocalDate getLocalDate() {
        return LocalDate.of(year, month, day);
    }
}

File: src/main/kotlin/com/qtech/utilities/datetime/Date.kt
package com.qtech.utilities.datetime

import java.io.Serializable
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.chrono.IsoChronology
import java.time.chrono.IsoEra
import java.util.*

class Date(var day: Int, var month: Int, var year: Int) : Serializable {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val date = other as Date
        return day == date.day && month == date.month && year == date.year
    }

    fun equalsIgnoreYear(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val date = o as Date
        return day == date.day &&
                month == date.month
    }

    override fun hashCode(): Int {
        return Objects.hash(day, month, year)
    }

    val dayOfWeek: DayOfWeek
        get() {
            val localDate = localDate
            return localDate.dayOfWeek
        }

    val dayOfYear: Int
        get() {
            val localDate = localDate
            return localDate.dayOfYear
        }

    val era: IsoEra
        get() {
            val localDate = localDate
            return localDate.era
        }

    val chronology: IsoChronology
        get() {
            val localDate = localDate
            return localDate.chronology
        }

    val isLeapYear: Boolean
        get() {
            val localDate = localDate
            localDate.toString()
            return localDate.chronology.isLeapYear(year.toLong())
        }

    private val localDate: LocalDate
        get() {
            return LocalDate.of(year, month, day)
        }

    companion object {
        fun current(): Date {
            val dateTime = LocalDateTime.now()
            val day = dateTime.dayOfMonth
            val month = dateTime.monthValue
            val year = dateTime.year
            return Date(day, month, year)
        }
    }
}